### PR TITLE
PUBDEV-5529 add leaf node assignment to DRF, GBM mojo

### PR DESCRIPTION
**Pull request by wendycwong**
1.  Added backend code to add leaf node assignment to DRF, GBM mojo
2.  Added Pyunit tests
3.  Added Runit tests

### CODE REVIEW

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoModel.java [hunk 1]
```diff
@@ -18,7 +18,7 @@
     private static final int NsdNaLeft = NaSplitDir.NALeft.value();
     private static final int NsdLeft = NaSplitDir.Left.value();
 
-    protected Number _mojo_version;
+    public Number _mojo_version;
```
Comment (michalkurka) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoModel.java]: Why is this needed? It shouldn't be, IMHO

File: h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java [hunk 2]
```diff
@@ -412,6 +416,7 @@ public BinomialModelPrediction predictBinomial(RowData data, double offset) thro
     double[] preds = preamble(ModelCategory.Binomial, data, offset);
 
     BinomialModelPrediction p = new BinomialModelPrediction();
+    p.leafNodeAssignments = leafNodeAssignment(data);  // assign leaf node assignment if desired
```
Comment (michalkurka) [inline h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java]: This should be optional, it will slow down the prediction.

I am thinking of having a separate API for this feature. We should not include it in the default response.

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoModel.java (hunk 1 above)
Comment (wendycwong) [reply h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoModel.java]: I need it to check the mojo version.  Leaf node assignment is only available after Pasha change. I don't need to check it at the outside.  Reverting it back now.

File: h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java (hunk 2 above)
Comment (wendycwong) [reply h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java]: Agreed.  It will only be executed if called separately and with the correct mojo-version.

Comment (wendycwong) [reply h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java]: I added a parameter to EasyPredictModelWrapper.java, boolean getLeafAssignment which is default to be false.  If a user want to get the leaf node assignment, they will need to call enableLeafAssignment() to set it to true.

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoModel.java [hunk 3]
```diff
@@ -30,15 +30,15 @@
      * multiclass response.
      */
     protected int _ntree_groups;
-    protected int _ntrees_per_group;
+    public int _ntrees_per_group;
```
Comment (michalkurka) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoModel.java]: @wendycwong can you please make these protected again? they are not used outside of this class

File: h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java [hunk 4]
```diff
@@ -56,6 +60,11 @@
   private final boolean convertUnknownCategoricalLevelsToNa;
   private final boolean convertInvalidNumbersToNa;
   private final boolean useExtendedOutput;
+  private boolean getLeafAssignment=false;
```
Comment (michalkurka) [inline h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java]: we should have this field final and use the Builder pattern that is used for the rest of the configuration properties, the config object should have the setters and the field should be final. What do you think?

File: h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java [hunk 5]
```diff
@@ -427,6 +438,15 @@ public BinomialModelPrediction predictBinomial(RowData data, double offset) thro
     return p;
   }
 
+  private String[] leafNodeAssignment(RowData data) throws PredictException {
```
Comment (michalkurka) [inline h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java]: I would even think about making this method public for the users who want finer grade control over the API.

File: h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java (hunk 4 above)
Comment (michalkurka) [reply h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java]: also in the constructor of EasyPredictModelWrapper we can check that leafnode assignment is only enabled when the users uses SharedTreeModel

otherwise they will fail at predict time on cast exception

**Comment by michalkurka** [comment]
@wendycwong thank you for incorporating my comments - I added a bunch of minor suggestions - please let me know what you think

File: h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoModel.java (hunk 3 above)
Comment (wendycwong) [reply h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoModel.java]: Done.  I forgot.  Sorry.

**Comment by wendycwong** [comment]
Michal:  they are good minor suggestions and I have incorporated them.  Thanks, Wendy

File: h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java [hunk 6]
```diff
@@ -56,6 +60,8 @@
   private final boolean convertUnknownCategoricalLevelsToNa;
   private final boolean convertInvalidNumbersToNa;
   private final boolean useExtendedOutput;
+  private final boolean getLeafAssignment;
```
Comment (michalkurka) [inline h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java]: enableLeafAssignment
